Code/Framework/AzFramework/AzFramework/Physics/DebugDraw/CharacterPhysicsDebugDraw.cpp

```diff
@@ -66,6 +108,105 @@ namespace Physics
+                        const AZ::Transform shape2Transform = node2Transform * AZ::Transform::CreateFromQuaternionAndTranslation(
+                            shapeColliderPair2.first->m_rotation, shapeColliderPair2.first->m_position);
+
+                        // check the collision layers
+                        bool shouldCollide = false;
+                        Physics::CollisionRequestBus::BroadcastResult(
+                            shouldCollide, &Physics::CollisionRequests::ShouldCollide, *shapeColliderPair1.first.get(),
+                            *shapeColliderPair2.first.get());
+
+                        if (!shouldCollide)
+                        {
+                            continue;
+                        }
+
+                        // test if the geometries of the colliders intersect
+                        using ShapeVariant = AZStd::variant<AZStd::monostate, AZ::Capsule, AZ::Obb, AZ::Sphere>;
+
+                        auto getShapeVariant = [](const Physics::ShapeConfiguration* shapeConfiguration,
+                                                  const AZ::Transform& shapeTransform) -> ShapeVariant
+                        {
+                            if (const auto* boxShapeConfiguration =
+                                    azdynamic_cast<const Physics::BoxShapeConfiguration*>(shapeConfiguration))
+                            {
+                                return boxShapeConfiguration->ToObb(shapeTransform);
+                            }
+                            else if (
+                                const auto* capsuleShapeConfiguration =
+                                    azdynamic_cast<const Physics::CapsuleShapeConfiguration*>(shapeConfiguration))
+                            {
+                                return capsuleShapeConfiguration->ToCapsule(shapeTransform);
+                            }
+                            else if (
+                                const auto* sphereShapeConfiguration =
+                                    azdynamic_cast<const Physics::SphereShapeConfiguration*>(shapeConfiguration))
+                            {
+                                return sphereShapeConfiguration->ToSphere(shapeTransform);
+                            }
+                            else
+                            {
+                                return AZStd::monostate{};
+                            }
+                        };
+
+                        ShapeVariant shapeVariant1 = getShapeVariant(shapeColliderPair1.second.get(), shape1Transform);
+                        ShapeVariant shapeVariant2 = getShapeVariant(shapeColliderPair2.second.get(), shape2Transform);
+                        bool overlaps = AZStd::visit(
+                            [](auto&& arg1, auto&& arg2)
+                            {
+                                return AZ::ShapeIntersection::Overlaps(arg1, arg2);
+                            },
```

Comment:
instead of stubbing out all those methods for monostate you can do this.
```
 using Arg1Type = AZStd::decay_t<decltype(arg1)>;
                                using Arg2Type = AZStd::decay_t<decltype(arg2)>;
                                if constexpr (AZStd::is_same_v<Arg1Type, AZStd::monostate> ||
                                    AZStd::is_same_v<Arg2Type, AZStd::monostate>) {
                                    return false;
                                } else {
                                    return AZ::ShapeIntersection::Overlaps(arg1, arg2);
                                }
```

Reply:
I tried something very similar originally but I couldn't get it to work. I'll try your version.

Reply:
+1 this would be a nice simplification if it works

Reply:
Yeah, your way works, nice!

I think what I tried before was identical to yours apart from I didn't use the else block, guess that's required for it to figure out at compile time that the following code can only be reached with non-monostate arguments.